corehq/apps/app_manager/management/commands/add_case_tile_template_field.py

```diff
@@ -0,0 +1,37 @@
+from corehq.apps.app_manager.management.commands.helpers import AppMigrationCommandBase
+from corehq.apps.app_manager.suite_xml.const import CASE_TILE_TEMPLATE_NAME_PERSON_SIMPLE
+from corehq.toggles import CASE_LIST_TILE
+
+
+class Command(AppMigrationCommandBase):
+    help = """
+    Changes the boolean Detail.use_case_tiles into a string to designate a template, Detail.case_tile_template.
+    """
+
+    include_builds = True
+    include_linked_apps = True
+    DOMAIN_LIST_FILENAME = 'add_case_tile_template_field.txt'
+    DOMAIN_PROGRESS_NUMBER_FILENAME = 'add_case_tile_template_field.txt'
+    chunk_size = 25
+
+    def get_domains(self):
+        return CASE_LIST_TILE.get_enabled_domains()
+
+    def migrate_app(self, app):
+        app_was_changed = False
+
+        def _migrate_detail(detail):
+            if detail and (detail.get('use_case_tiles') is not None):
+                if detail['use_case_tiles']:
+                    detail['case_tile_template'] = CASE_TILE_TEMPLATE_NAME_PERSON_SIMPLE
+                del detail['use_case_tiles']
+                return True
+            return False
+
+        for module in app['modules']:
+            detail_pair = module.get('case_details')
+            if detail_pair:
+                short_detail, long_detail = detail_pair['short'], detail_pair['long']
+                app_was_changed = app_was_changed or _migrate_detail(short_detail) or _migrate_detail(long_detail)
```

Comment:
`or` evaluates lazily, which means that once `app_was_changed` is set to `True`, `_migrate_detail` won't be called for the remainder of the app

```python
In [1]: def migrate(val):
   ...:     print(f"migrating {val}")
   ...:     return True
   ...: 

In [2]: False or migrate(1) or migrate(2)
migrating 1
Out[2]: True

In [3]: True or migrate(3) or migrate(4)
Out[3]: True
```
You could keep it as a oneliner with something like this:
```python
app_was_changed = any([app_was_changed, _migrate_detail(short_detail), _migrate_detail(long_detail)])```
or use intermediate variables to store the results of the short and long detail migrations.

Reply:
also, FWIW `iter_update` compares the state of the doc before and after wrapping and only saves it if there was a change:
https://github.com/dimagi/commcare-hq/blob/0a602cdc5d324507a3b5040d6309d9b80a5af3c2/corehq/util/couch.py#L244-L249
Though with apps as complicated as they are, I wouldn't be surprised if there are some innocuous changes that happen with wrapping docs that we don't need to actually commit to the DB, so being more deliberate about it seems reasonable.

Reply:
A list literal (written with `[` and `]`) is evaluated immediately, so for:
```python
any([1, foo()])
```
The first step would be to evaluate the list, holding it in memory, then that list is passed to the `any` function, which will only need to inspect the first element of the list before returning `True`.  However, the whole list was evaluated before being passed to `any` regardless, which is why the interpreter outputs `foo` in your example above.

In my example, the generator is coerced to a list (`list(myitems())`), which forces it to be fully evaluated at that point.  As another example, the `range` function was turned into a generator sort of thing in Python 3, so here's a nice demonstration of the difference:
```python
In [1]: import timeit
   ...: def lazy():
   ...:     a = any(range(1_000_000))
   ...: def greedy():
   ...:     a = any(list(range(1_000_000)))
   ...: 

In [2]: timeit.timeit(lazy, number=100)
Out[2]: 1.1786993127316236e-05

In [3]: timeit.timeit(greedy, number=100)
Out[3]: 1.9306027129932772
```